File: sdk/ruby/test/test_protocol_fixtures.rb
# frozen_string_literal: true

require 'minitest/autorun'
require 'json'
require_relative '../lib/mlld'

class ProtocolFixturesTest < Minitest::Test
  def setup
    @client = Mlld::Client.new
  end

  def test_execute_result_fixture_preserves_security
    fixture = load_fixture('execute-result.json')
    result = @client.send(:decode_execute_result, fixture.fetch('result'), [])

    assert_equal(1, result.state_writes.length)
    assert_equal(1, result.sessions.length)
    assert_equal('planner', result.sessions.first.name)
    assert_equal({ 'count' => 2, 'status' => 'done' }, result.sessions.first.final_state)
    assert_equal(['trusted'], result.state_writes.first.security['labels'])
    assert_equal(['trusted'], result.effects.first.security['labels'])
  end

  def test_analyze_result_fixture_uses_trigger
    fixture = load_fixture('analyze-result.json')
    result = @client.send(:build_analyze_result, fixture.fetch('result'), 'fallback.mld')

    assert_equal(2, result.guards.length)
    assert_equal('secret', result.guards.first.trigger)
    assert_equal('', result.guards.last.name)
    assert_equal('net:w', result.guards.last.trigger)
  end

  def test_state_write_event_fixture_preserves_security
    fixture = load_fixture('state-write-event.json')
    state_write = @client.send(:state_write_from_event, fixture.fetch('event'))

    refute_nil(state_write)
    assert_equal('payload', state_write.path)
    assert_equal(['trusted'], state_write.security['labels'])
  end

  def test_session_write_event_fixture_preserves_fields
    fixture = load_fixture('session-write-event.json')
    session_write = @client.send(:session_write_from_event, fixture.fetch('event'))

    refute_nil(session_write)
    assert_equal('planner', session_write.session_name)
    assert_equal('count', session_write.slot_path)
    assert_equal('increment', session_write.operation)
    assert_equal(1, session_write.prev)
    assert_equal(2, session_write.next)
  end

  def test_trace_event_fixture_preserves_fields
    fixture = load_fixture('trace-event.json')
    trace_event = @client.send(:trace_event_from_event, fixture.fetch('event'))

    refute_nil(trace_event)
    assert_equal('guard.deny', trace_event.event)
    assert_equal('guard', trace_event.category)
    assert_equal('frame-parent', trace_event.scope['parentFrameId'])
    assert_equal('send', trace_event.data['operation'])
  end

  def test_error_fixture_decodes_transport_error
    fixture = load_fixture('error-result.json')
    error = @client.send(:error_from_payload, fixture.fetch('error'))

    assert_instance_of(Mlld::Error, error)
    assert_equal('TIMEOUT', error.code)
    assert_match(/timeout/i, error.message)
  end

  def test_sign_result_fixture_decodes_file_verify_result
    fixture = load_fixture('sign-result.json')
    result = @client.send(:file_verify_result_from_payload, fixture.fetch('result'))

    assert_equal('docs/a.txt', result.relative_path)
    assert_equal('sha256:abc', result.expected_hash)
    assert_equal({ 'purpose' => 'sdk' }, result.metadata)
  end

  def test_fs_status_fixture_decodes_filesystem_status
    fixture = load_fixture('fs-status-result.json')
    result = @client.send(:filesystem_status_from_payload, fixture.fetch('result').first)

    assert_equal('docs/a.txt', result.relative_path)
    assert_equal(['trusted'], result.labels)
    assert_equal(['secret'], result.taint)
  end

  def test_sign_content_fixture_decodes_content_signature
    fixture = load_fixture('sign-content-result.json')
    result = @client.send(:content_signature_from_payload, fixture.fetch('result'))

    assert_equal('content-1', result.id)
    assert_equal('user:alice', result.signed_by)
    assert_equal({ 'channel' => 'sdk' }, result.metadata)
  end

  def test_process_and_execute_requests_serialize_trace_memory
    process_params = @client.build_process_request(
      'show "hi"',
      trace: 'effects',
      trace_memory: true,
      trace_file: 'trace.jsonl',
      trace_stderr: false
    )

    assert_equal 'effects', process_params['trace']
    assert_equal true, process_params['traceMemory']
    assert_equal 'trace.jsonl', process_params['traceFile']
    assert_equal false, process_params['traceStderr']

    execute_params = @client.build_execute_request(
      '/repo/main.mld',
      { 'name' => 'Ada' },
      trace_memory: true
    )

    assert_equal true, execute_params['traceMemory']
  end

  def test_transport_command_adds_runtime_heap_args
    wrapper = Mlld::Client.new(heap: '8g', heap_snapshot_near_limit: 2)
    assert_equal(
      ['mlld', '--mlld-heap=8g', '--heap-snapshot-near-limit', '2', 'live', '--stdio'],
      wrapper.send(:transport_command)
    )

    node = Mlld::Client.new(command: 'node', command_args: ['./dist/cli.cjs'], heap: '8g', heap_snapshot_near_limit: 2)
    assert_equal(
      ['node', '--max-old-space-size=8192', '--heapsnapshot-near-heap-limit=2', './dist/cli.cjs', 'live', '--stdio'],
      node.send(:transport_command)
    )

    assert_raises(Mlld::Error) { Mlld::Client.new(command: 'node', heap: 'nope').send(:transport_command) }
    assert_raises(Mlld::Error) { Mlld::Client.new(heap_snapshot_near_limit: 0).send(:transport_command) }
  end

  private

  def load_fixture(name)
    path = File.expand_path(File.join('..', '..', 'fixtures', name), __dir__)
    JSON.parse(File.read(path))
  end
end

File: sdk/python/test_protocol_fixtures.py
from __future__ import annotations

import json
import unittest
from pathlib import Path

import mlld


FIXTURE_DIR = Path(__file__).resolve().parents[1] / "fixtures"


def _load_fixture(name: str) -> dict[str, object]:
    return json.loads((FIXTURE_DIR / name).read_text())


class ProtocolFixturesTest(unittest.TestCase):
    def test_execute_result_fixture_round_trips_all_fields(self) -> None:
        envelope = _load_fixture("execute-result.json")
        self.assertEqual(set(envelope.keys()), {"id", "result"})

        result = mlld._execute_result_from_payload(envelope["result"], [], [])

        self.assertEqual(result.output, "fixture-output")
        self.assertEqual(result.state_writes[0].path, "payload")
        self.assertEqual(result.state_writes[0].value, {"enabled": True, "nested": {"count": 2}})
        self.assertEqual(result.sessions[0].name, "planner")
        self.assertEqual(result.sessions[0].final_state, {"count": 2, "status": "done"})
        self.assertEqual(
            result.state_writes[0].security,
            {"labels": ["trusted"], "taint": ["untrusted"]},
        )
        self.assertEqual(result.effects[0].security, {"labels": ["trusted"]})
        self.assertEqual(result.denials[0].operation, "send")
        self.assertEqual(result.metrics.total_ms, 5.5)

    def test_analyze_result_fixture_round_trips_guard_trigger(self) -> None:
        envelope = _load_fixture("analyze-result.json")
        client = mlld.Client(timeout=1.0)

        def fake_request(method: str, params: dict[str, object], timeout: float | None):
            self.assertEqual(method, "analyze")
            self.assertEqual(params, {"filepath": "/repo/routes/example.mld"})
            self.assertIsNone(timeout)
            return envelope["result"], [], []

        client._request = fake_request  # type: ignore[method-assign]

        result = client.analyze("/repo/routes/example.mld")

        self.assertEqual(result.guards[0].trigger, "secret")
        self.assertEqual(result.guards[1].trigger, "net:w")
        self.assertEqual(result.guards[1].name, "")

    def test_state_write_event_fixture_round_trips_security(self) -> None:
        envelope = _load_fixture("state-write-event.json")

        event = mlld._state_write_from_event(envelope["event"])

        self.assertIsNotNone(event)
        assert event is not None
        self.assertEqual(event.path, "payload")
        self.assertEqual(event.value, {"enabled": True})
        self.assertEqual(event.security, {"labels": ["trusted"], "taint": ["secret"]})

    def test_session_write_event_fixture_round_trips_fields(self) -> None:
        envelope = _load_fixture("session-write-event.json")

        event = mlld._session_write_from_event(envelope["event"])

        self.assertIsNotNone(event)
        assert event is not None
        self.assertEqual(event.session_name, "planner")
        self.assertEqual(event.slot_path, "count")
        self.assertEqual(event.operation, "increment")
        self.assertEqual(event.prev, 1)
        self.assertEqual(event.next, 2)

    def test_guard_denial_event_fixture_round_trips_fields(self) -> None:
        envelope = _load_fixture("guard-denial-event.json")

        event = mlld._guard_denial_from_event(envelope["event"])

        self.assertIsNotNone(event)
        assert event is not None
        self.assertEqual(event.guard, "@blocker")
        self.assertEqual(event.operation, "send")
        self.assertEqual(event.args, {"value": "hello"})

    def test_trace_event_fixture_round_trips_fields(self) -> None:
        envelope = _load_fixture("trace-event.json")

        event = mlld._trace_event_from_event(envelope["event"])

        self.assertIsNotNone(event)
        assert event is not None
        self.assertEqual(event.event, "guard.deny")
        self.assertEqual(event.category, "guard")
        self.assertEqual(event.scope["parentFrameId"], "frame-parent")
        self.assertEqual(event.data["operation"], "send")

    def test_fs_status_fixture_round_trips_array_payload(self) -> None:
        envelope = _load_fixture("fs-status-result.json")
        client = mlld.Client(timeout=1.0)

        def fake_request(method: str, params: dict[str, object], timeout: float | None):
            self.assertEqual(method, "fs:status")
            self.assertEqual(params, {"glob": "docs/*.txt", "basePath": "/repo"})
            self.assertEqual(timeout, 5.0)
            return envelope["result"], [], []

        client._request = fake_request  # type: ignore[method-assign]

        entries = client.fs_status("docs/*.txt", base_path="/repo", timeout=5.0)

        self.assertEqual(len(entries), 1)
        self.assertEqual(entries[0].relative_path, "docs/a.txt")
        self.assertEqual(entries[0].labels, ["trusted"])
        self.assertEqual(entries[0].taint, ["secret"])

    def test_sign_result_fixture_round_trips_object_payload(self) -> None:
        envelope = _load_fixture("sign-result.json")
        client = mlld.Client(timeout=1.0)

        def fake_request(method: str, params: dict[str, object], timeout: float | None):
            self.assertEqual(method, "sig:sign")
            return envelope["result"], [], []

        client._request = fake_request  # type: ignore[method-assign]

        result = client.sign("docs/a.txt")

        self.assertEqual(result.relative_path, "docs/a.txt")
        self.assertEqual(result.metadata, {"purpose": "sdk"})

    def test_sign_content_fixture_round_trips_object_payload(self) -> None:
        envelope = _load_fixture("sign-content-result.json")
        client = mlld.Client(timeout=1.0)

        def fake_request(method: str, params: dict[str, object], timeout: float | None):
            self.assertEqual(method, "sig:sign-content")
            return envelope["result"], [], []

        client._request = fake_request  # type: ignore[method-assign]

        result = client.sign_content("hello world", "user:alice")

        self.assertEqual(result.id, "content-1")
        self.assertEqual(result.signed_by, "user:alice")
        self.assertEqual(result.metadata, {"channel": "sdk"})

    def test_error_fixture_round_trips_transport_error_shape(self) -> None:
        envelope = _load_fixture("error-result.json")
        self.assertEqual(set(envelope.keys()), {"id", "error"})

        error = mlld._error_from_payload(envelope["error"])

        self.assertEqual(error.code, "TIMEOUT")
        self.assertEqual(error.message, "request timeout after 5s")


if __name__ == "__main__":
    unittest.main()
